xdsl/parser/attribute_parser.py

```diff
@@ -643,20 +645,98 @@ def _parse_optional_builtin_parametrized_attr(self) -> Attribute | None:
+            self.raise_error("Hex string in denseAttr is invalid")
+
+        # Use struct builtin package for unpacking f32, f64
+        format_str: str = ""
+        match element_type:
+            case Float32Type():
+                chunk_size = 4
+                format_str = "@f"  # @ in format string implies native endianess
+            case Float64Type():
+                chunk_size = 8
+                format_str = "@d"
+            case IntegerType():
+                if element_type.width.data % 8 != 0:
+                    self.raise_error(
+                        "Hex strings for dense literals only support integer types that are a multiple of 8 bits"
+                    )
+                chunk_size = element_type.width.data // 8
+            case _:
+                self.raise_error(
+                    "Hex strings for dense literals are only supported for int, f32 and f64 types"
+                )
+        num_chunks = len(byte_list) // chunk_size
+
+        data_values: list[int] | list[float] = []
+
+        if isa(element_type, Float32Type | Float64Type):
+            for i in range(num_chunks):
+                parsed_float = struct.unpack(
+                    format_str, byte_list[i * chunk_size : (i + 1) * chunk_size]
+                )
+                data_values.append(*parsed_float)
+        # Use int for unpacking IntegerType
+        else:
+            for i in range(num_chunks):
+                parsed_int = int.from_bytes(
+                    byte_list[i * chunk_size : (i + 1) * chunk_size],
+                    sys.byteorder,
+                    signed=True,
+                )
+                data_values.append(parsed_int)
+        # Splat attribute case, same value everywhere
+        if chunk_size == len(byte_list) and type.get_shape() != (1,):
+            return data_values, []
+        return data_values, [num_chunks]
+
     def _parse_builtin_dense_attr(self, _name: Span) -> DenseIntOrFPElementsAttr:
         self.parse_punctuation("<", " in dense attribute")
 
         # The flatten list of elements
-        values: list[AttrParser._TensorLiteralElement]
+        values: list[AttrParser._TensorLiteralElement] | list[int] | list[float]
         # The dense shape.
         # If it is `None`, then there is no values.
         # If it is `[]`, then this is a splat attribute, meaning it has the same
         # value everywhere.
         shape: list[int] | None
+        hex_string: str | None = None
         if self._current_token.text == ">":
             values, shape = [], None
         else:
-            values, shape = self._parse_tensor_literal()
+            hex_string = self.parse_optional_str_literal()
+            if hex_string is not None:
+                # Can not determine values without type yet
+                values, shape = [], None
+            else:
+                # Expect a tensor literal instead
+                values, shape = self._parse_tensor_literal()
```

Comment:
I have trouble understanding what is expected to be at each line here.

Can this be simplified for the case where `values, shape = [], None`?

Can this be combined with the `_parse_builtin_dense_attr_hex` below?
You seem to assume this is a hex, but you don't really know that yet, wouldn't determining this as fast as possible make sense and potentially elide the extra check `if hex_string is not None:` below?

Ideally, I'd like to come out of the block knowing that if hex exists is valid.

Reply:
Hi @compor I agree, that this is a bit of an ugly way of going about this.

I think it looks ugly because of the following problems I had:

* The problem is that I can not really extract the shape information out of the hex_string since the information is already flattened inside the hex attribute, so there are a few checks you can not perform on hex values.
* The added code kind of breaks the previous flow because you already need your type info to be able to parse the hex string in the first place. In the previous code this was not a problem, because you can already parse the ints and the floats purely on the format of the numbers
* another problem is that I was afraid of emitting `_TensorLiteralElement`s because they seem to require a span, and I was unsure on how this works. @superlopuh would this solve your question as well?

Does this make sense? Thanks for your help!

Reply:
To answer your actual question @compor:

Originally:
`shape = []` for splat attribute, list of ints for bigger attributes, e.g. [[2,3],[3,2]] would give a `shape = [2,2] `
`values` = a list of `_TensorLiteralElement`s which contains parsed `TensorLiteral` elements

Now:
`hex_string` The value of the hex string, if present, `None` if not present
`shape = []` for splat attribute, list of ints for bigger attributes, e.g. [[2,3],[3,2]] would give a `shape = [2,2] `.
For hex attributes I can only give the flattened shape as a single-element list (e.g. `shape = [4]`) because of a lack of information.
`values` = a list of `_TensorLiteralElement`s which contains parsed `TensorLiteral` elements
In the case of a hex attribute, the `values` are already in their python counterparts, so values is already of type `list[int]` or `list[float]` and doesn't need to be converted from TensorLiteralElement anymore.

I agree that this is a lot of info necessary for just 3 variables :sweat_smile: , any suggestions very much welcome!

Reply:
I had a similar issue with this BTW, when I tried to implement `memref.global` custom syntax: https://xdsl.zulipchat.com/#narrow/stream/368604-Framework/topic/printing.20and.20parsing.20attributes/near/410844828

Reply:
We could just put the entire span for the string, so we don't have `list[int]` or `list[float]` and have the list of literal elements, I would be fine with that!
We won't use these errors anyway, so there is no problems.

Reply:
To simplify the logic actually, can we just split the following code in two functions? One function to deal with the hex, and the other to deal with the lists?
Because currently the logic gets quite complex

Reply:
Thanks @math-fehr I'll try to refactor to using the literals now and I'll see what I'm left with :smile:

Reply:
On second thought @math-fehr it doesn't make a lot of sense to make the _TensorLiteralElements, because then you have to convert them to a list of ints going out of the function anyways :(